### PR TITLE
Update to latest cadvisor - cleanup mesos/rkt

### PR DESCRIPTION
**Pull request by dims**
Change-Id: I2aa91b4dd1f421fdf72623e5dc94d4f304d9b76b

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Mainly to bring in the code from the following PRs:
- https://github.com/google/cadvisor/pull/2209
- https://github.com/google/cadvisor/pull/2213

This cleans up rkt/mesos and their dependencies

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 1]
```diff
@@ -19,20 +19,33 @@ limitations under the License.
...
 	"net/http"
 	"os"
 	"path"
 	"time"
 
+	"github.com/google/cadvisor/accelerators"
 	"github.com/google/cadvisor/cache/memory"
+	"github.com/google/cadvisor/container"
 	cadvisormetrics "github.com/google/cadvisor/container"
+	_ "github.com/google/cadvisor/container/containerd"
```
Comment (liggitt) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: can you split the kubernetes changes into a separate commit than the `hack/update-vendor.sh` commit?

**Comment by liggitt** [comment]
can you make sure `hack/lint-dependencies.sh` doesn't show new unneeded replace directives?

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 1 above)
Comment (dims) [reply pkg/kubelet/cadvisor/cadvisor_linux.go]: ack. on it

**Comment by dims** [comment]
@liggitt - i ran `hack/lint-dependencies.sh | grep pin-dep | bash` until it stopped updating anything

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 2]
```diff
@@ -113,6 +127,122 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, usingLegacySt
...
+	if memoryCache == nil {
+		return nil, fmt.Errorf("manager requires memory storage")
+	}
+
+	// Detect the container we are running on.
+	selfContainer, err := cgroups.GetOwnCgroupPath("cpu")
+	if err != nil {
+		return nil, err
+	}
+	klog.V(2).Infof("cAdvisor running in container: %q", selfContainer)
```
Comment (tallclair) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: nit: s/cadvisor/kubelet/

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 3]
```diff
@@ -113,6 +127,122 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, usingLegacySt
...
+
+	var (
+		dockerStatus info.DockerStatus
+		rktPath      string
```
Comment (tallclair) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: unused

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 4]
```diff
@@ -113,6 +127,122 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, usingLegacySt
...
+		klog.V(5).Infof("CRI-O not connected: %v", err)
+	}
+
+	context := fs.Context{
+		Docker: fs.DockerContext{
+			Root:         docker.RootDir(),
+			Driver:       dockerStatus.Driver,
+			DriverStatus: dockerStatus.DriverStatus,
+		},
+		RktPath: rktPath,
```
Comment (tallclair) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: remove this

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 5]
```diff
@@ -113,6 +127,122 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, usingLegacySt
...
+		return nil, err
+	}
+	klog.V(1).Infof("Machine: %+v", *machineInfo)
+
+	newManager := manager.New(
+		memoryCache,
+		fsInfo,
+		sysfs,
+		*machineInfo,
+		make([]chan error, 0, 2),
```
Comment (tallclair) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: nit: use a named variable

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 6]
```diff
@@ -113,6 +127,122 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, usingLegacySt
...
+
+	versionInfo, err := newManager.GetVersionInfo()
+	if err != nil {
+		return nil, err
+	}
+	klog.V(1).Infof("Version: %+v", *versionInfo)
+	return newManager, nil
+}
+
+func retryDockerStatus() info.DockerStatus {
```
Comment (tallclair) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: There's a lot of code here copied from cadvisor. I'd prefer to see more of this exposed as methods for reuse, but I don't know that we need to block the PR on it.

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 7]
```diff
@@ -113,6 +127,122 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, usingLegacySt
...
+	dockerStatus = retryDockerStatus()
+
+	crioClient, err := crio.Client()
+	if err != nil {
+		return nil, err
+	}
+	crioInfo, err := crioClient.Info()
+	if err != nil {
+		klog.V(5).Infof("CRI-O not connected: %v", err)
+	}
```
Comment (tallclair) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: @dashpole @yujuhong @derekwaynecarr @mrunalp 

Per [this comment](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cadvisor/util.go#L66-L72):
```
// CRI integrations should get container metrics via CRI. Docker
// uses the built-in cadvisor to gather such metrics on Linux for
// historical reasons.
// cri-o relies on cadvisor as a temporary workaround. The code should
// be removed. Related issue:
// https://github.com/kubernetes/kubernetes/issues/51798
// UsingLegacyCadvisorStats returns true if container stats are provided by cadvisor instead of through the CRI
```

Gathering crio stats from cadvisor is a temporary workaround. The referenced issue was auto-closed. What are the conditions for removing the cri-o dependency here?

@dims - If we can't remove it now, can you file a new issue with the criteria identified, and add a comment here to remove this once that issue is resolved.

**Review by tallclair — COMMENTED** [review]
Thanks!

I added a release note, per @dashpole 's [comment](https://github.com/google/cadvisor/pull/2213#issuecomment-480048462).

This PR is marked WIP, what is the remaining work?

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 7 above)
Comment (dashpole) [reply pkg/kubelet/cadvisor/cadvisor_linux.go]: @mrunalp can correct me, but I don't think CRI-O implements the stats interface yet.

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 6 above)
Comment (dims) [reply pkg/kubelet/cadvisor/cadvisor_linux.go]: thanks!

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 7 above)
Comment (tallclair) [reply pkg/kubelet/cadvisor/cadvisor_linux.go]: From the code it looks like support is there: https://github.com/cri-o/cri-o/blob/master/server/container_stats_list.go

Comment (dims) [reply pkg/kubelet/cadvisor/cadvisor_linux.go]: So i have reopened https://github.com/kubernetes/kubernetes/issues/51798

and will add a comment above `crioInfo, err := crioClient.Info()` similar to what's above `UsingLegacyCadvisorStats` method

**Review by cblecker — APPROVED** [review]
Explicit sign off for the go module/vendor/license changes. Haven't reviewed the specific cadvisor code changes to the kubelet

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 3 above)
Comment (dims) [reply pkg/kubelet/cadvisor/cadvisor_linux.go]: fixed

File: pkg/kubelet/cadvisor/cadvisor_linux.go [hunk 8]
```diff
@@ -89,7 +103,7 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, usingLegacySt
 	// collect metrics for all cgroups
 	rawContainerCgroupPathPrefixWhiteList := []string{"/"}
 	// Create and start the cAdvisor container manager.
-	m, err := manager.New(memory.New(statsCacheDuration, nil), sysFs, maxHousekeepingInterval, allowDynamicHousekeeping, includedMetrics, http.DefaultClient, rawContainerCgroupPathPrefixWhiteList)
+	m, err := newCadvisor(memory.New(statsCacheDuration, nil), sysFs, maxHousekeepingInterval, allowDynamicHousekeeping, includedMetrics, http.DefaultClient, rawContainerCgroupPathPrefixWhiteList)
```
Comment (liggitt) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: talked with @dims about an approach that would avoid needing to change this signature and keep us (and all callers) from having to copy out initialization code

will open a PR to cadvisor and sync up there

Comment (liggitt) [inline pkg/kubelet/cadvisor/cadvisor_linux.go]: talked with @dims about an approach that would avoid needing to change this signature and keep us (and all callers) from having to copy out initialization code

will open a PR to cadvisor and sync up there

**Comment by liggitt** [comment]
/hold

per https://github.com/kubernetes/kubernetes/pull/76291/commits/80db806887099aa8447df0ee94b55ecd610d51fd#r273689574

**Comment by dims** [comment]
Thanks @liggitt

File: pkg/kubelet/cadvisor/cadvisor_linux.go (hunk 8 above)
Comment (liggitt) [reply pkg/kubelet/cadvisor/cadvisor_linux.go]: opened https://github.com/google/cadvisor/pull/2217, keeping all the collector initialization contained to `manager.New()` ... testing with CI in https://github.com/kubernetes/kubernetes/pull/76355
